blockchain/state_migration.go

```diff
@@ -641,26 +624,89 @@ func printDepthStats(depthCounter map[int]int) {
 	}
 }
 
-var errNotExistingAddress = fmt.Errorf("there is no account corresponding to the given address")
-var errNotContractAddress = fmt.Errorf("given address is not a contract address")
-
 // GetContractStorageRoot returns the storage root of a contract based on the given block.
 func (bc *BlockChain) GetContractStorageRoot(block *types.Block, db state.Database, contractAddr common.Address) (common.Hash, error) {
 	stateDB, err := state.New(block.Root(), db)
 	if err != nil {
 		return common.Hash{}, fmt.Errorf("failed to get StateDB - %w", err)
 	}
+	return stateDB.GetContractStorageRoot(contractAddr)
+}
 
-	acc := stateDB.GetAccount(contractAddr)
-	if acc == nil {
-		return common.Hash{}, errNotExistingAddress
+// prepareWarmUp creates and returns resources needed for state warm-up.
+func (bc *BlockChain) prepareWarmUp() (*types.Block, state.Database, *statedb.Database, error) {
+	// There is a chance of concurrent access to quitWarmUp, though not likely to happen.
+	if bc.quitWarmUp != nil {
+		return nil, nil, nil, fmt.Errorf("already warming up")
 	}
-	if acc.Type() != account.SmartContractAccountType {
-		return common.Hash{}, errNotContractAddress
+
+	block := bc.GetBlockByNumber(bc.lastCommittedBlock)
+	if block == nil {
+		return nil, nil, nil, fmt.Errorf("block #%d not found", bc.lastCommittedBlock)
+	}
+
+	mainTrieDB := bc.StateCache().TrieDB()
+	cache := mainTrieDB.TrieNodeCache()
+	if cache == nil {
+		return nil, nil, nil, fmt.Errorf("target cache is nil")
 	}
-	contract, true := acc.(*account.SmartContractAccount)
-	if !true {
-		return common.Hash{}, errNotContractAddress
+	db := state.NewDatabaseWithExistingCache(bc.db, cache)
+	return block, db, mainTrieDB, nil
+}
+
+// runStateDBIterator runs the given statedb.Iterator until it reaches end.
+// If it reaches end, it will send a nil error to errCh to indicate that it has been finished.
+func runStateDBIterator(itr *statedb.Iterator, bc *BlockChain, resultCh chan struct{}, errCh chan error) (resultErr error) {
+	defer func() { errCh <- resultErr }()
+
+	for itr.Next() {
+		resultCh <- struct{}{}
+		select {
+		case <-bc.quitWarmUp:
+			return stopWarmUpErr
+		case <-bc.quit:
+			return blockChainStopWarmUpErr
+		default:
+		}
 	}
-	return contract.GetStorageRoot(), nil
+	return nil
+}
+
+// StartContractWarmUp retrieves a storage trie of the latest state root and caches the trie
+// corresponding to the given contract address.
+func (bc *BlockChain) StartContractWarmUp(contractAddr common.Address) error {
+	block, db, mainTrieDB, err := bc.prepareWarmUp()
+	if err != nil {
+		return err
+	}
+
+	stateDB, err := state.New(block.Root(), db)
+	if err != nil {
+		return fmt.Errorf("failed to get StateDB, err: %w", err)
+	}
+	storageTrieRoot, err := stateDB.GetContractStorageRoot(contractAddr)
+	if err != nil {
+		return err
+	}
+	itrs, err := stateDB.NewStorageTrieIterators(storageTrieRoot)
```

Comment:
How about calling this line in `runStateDBIterator` to match the codes of `StartWarmUp`?

Reply:
> LGTM :)
> But, why did we closed #810?
> This PR contains duplicated codes.

@winnie-byun When I closed #810, I thought this PR would be quite different from #810 because of `NewStorageTrieIterators`. But it is done easier and simpler than I expected. 

> How about calling this line in `runStateDBIterator` to match the codes of `StartWarmUp`?

I updated the codes to look as similar as possible, could you please check https://github.com/klaytn/klaytn/pull/831/commits/b7880c8f8a92e431f7755163be743c667f8c1192 ?

Reply:
FYI

## StartWarmUp
```
// StartWarmUp retrieves all state/storage tries of the latest state root and caches the tries.
func (bc *BlockChain) StartWarmUp() error {
	block, db, mainTrieDB, err := bc.prepareWarmUp()
	if err != nil {
		return err
	}
	// retrieve children nodes of state trie root node
	children, err := db.TrieDB().NodeChildren(block.Root())
	if err != nil {
		return err
	}
	// run goroutine for each child node
	resultCh := make(chan struct{}, 10000)
	errCh := make(chan error)
	bc.quitWarmUp = make(chan struct{})
	for _, child := range children {
		go bc.iterateStateTrie(child, db, resultCh, errCh)
	}
	// run a warm-up checker routine
	go bc.warmUpChecker(mainTrieDB, len(children), resultCh, errCh)
	logger.Info("State trie warm-up is started", "blockNum", block.NumberU64(),
		"root", block.Root().String(), "len(children)", len(children))
	return nil
}
```

## StartContractWarmUp
```
// StartContractWarmUp retrieves a storage trie of the latest state root and caches the trie
// corresponding to the given contract address.
func (bc *BlockChain) StartContractWarmUp(contractAddr common.Address) error {
	block, db, mainTrieDB, err := bc.prepareWarmUp()
	if err != nil {
		return err
	}
	// prepare contract storage trie specific resources - storageTrieRoot and storageTrie
	storageTrieRoot, storageTrie, err := prepareContractWarmUp(block, db, contractAddr)
	if err != nil {
		return fmt.Errorf("failed to prepare contract warm-up, err: %w", err)
	}
	// retrieve children nodes of contract storage trie root node
	children, err := db.TrieDB().NodeChildren(storageTrieRoot)
	if err != nil {
		return err
	}
	// run goroutine for each child node
	resultCh := make(chan struct{}, 10000)
	errCh := make(chan error)
	bc.quitWarmUp = make(chan struct{})
	for _, child := range children {
		go bc.iterateStorageTrie(child, storageTrie, resultCh, errCh)
	}
	// run a warm-up checker routine
	go bc.warmUpChecker(mainTrieDB, len(children), resultCh, errCh)
	logger.Info("Contract storage trie warm-up is started",
		"blockNum", block.NumberU64(), "root", block.Root().String(), "contractAddr", contractAddr.String(),
		"contractStorageRoot", storageTrieRoot.String(), "len(children)", len(children))
	return nil
}
```